docs/advanced/debugging.rst

```diff
@@ -27,15 +27,335 @@ And, of course, if you're hitting that ever-challenging ``No scope with a Tag ma
+* Both the ``IService1`` and ``IService2`` need an ``IService3`` and a new one was created for each instance.
+* ``IService3`` is decorated - see how it chains down to a node that looks more like a box? That indicates there's a decorator in play. You can see the component (the decorator) and the target (the thing being decorated) listed in that box.
+* ``IService3`` takes in an ``ILifetimeScope`` constructor parameter.
+
+That last one - the ``ILifetimeScope`` parameter - means ``IService3`` might do some service location (manually resolve something) inside the code. If you really want to know the full chain, you might have to correlate this graph with others. But how?
+
+Notice at the top there's an ``Operation #1`` counter - this counter increments for every resolve operation passing through the tracer. You can look for traces that have a larger number and do a little manual correlation. Unfortunately, this is the best we can offer because each resolution is entirely independent - service location breaks the chain. You can't assume all resolves associated with a lifetime scope are correlated because, for example, maybe all the resolutions for an entire application come from a single scope.
+
+Errors will also be highlighted so you can see where the error occurs.
+
+.. figure:: media/DotErrorTrace.png
+    :align: center
+    :alt: DOT graph for an error during a resolve.
+
+    DOT graph for an error during a resolve.
+
+In this case, you can see the trace all the way down to the location where it failed is highlighted red and bold. You can also see the exception type and message.
+
+Custom Tracers
+==============
+
+Using ``System.Diagnostics.DiagnosticSource`` in Autofac allows you to create your own custom tracer that can handle various events and generate the data that interests you.
+
+Events in the overall pipeline happen in this order:
+
+* Operation start
+    * Resolve request start
+        * Middleware start
+        * Middlware success / fail
+    * Resolve request success / fail
+* Operation success / fail
+
+Middleware may start additional resolve requests; and there's more than one middleware item in the pipeline. You can dig deeper on this in the :doc:`Pipelines <pipelines>` page.
+
+If you want to trace an *entire operation* from end to end, similar to the ``DefaultDiagnosticTracer``, you can start with the ``Autofac.Diagnostics.OperationDiagnosticTracerBase<TContent>`` class. This is the class on which the ``DefaultDiagnosticTracer`` is based. It is intentionally wired up to listen to all resolve events, from start to end, and track a full operation at a time. Your best example of this is to look at the source code for ``DefaultDiagnosticTracer``. Given there are a lot of events to handle, there's a lot of data to capture.
```

Comment:
Perhaps put a link in to the actual `DefaultDiagnosticTracer` source file on develop if you're suggesting people go take a look?

Also, line breaks...

Reply:
I wanted the docs to be refactor-safe, so I guess I could put a permalink to a specific version of the commit, but I didn't want to put it to a specific breakable link in a branch where things could move over time.

Reply:
Linked to the most recent commit of `DefaultDiagnosticTracer`.